File: src/main/java/com/example/indentory_management_system/Service/ProductService.java
package com.example.indentory_management_system.Service;

import java.util.List;

import com.example.indentory_management_system.dto.*;

public interface ProductService {

    public ProductResponsedto createProduct(ProductRequestdto productRequestDto);
    public List<ProductResponsedto> getAllProducts();
    public List<ProductResponsedto> getActiveProducts();
    public ProductResponsedto getByProductId(Long id);
    public ProductResponsedto getByProductSku(String sku);
    public ProductResponsedto getByProductName(String name);
    public ProductResponsedto updateProduct(Long id, ProductRequestdto productRequestDto);
    public ProductResponsedto deleteProduct(Long id);
    public List<ProductResponsedto> findproductCategory(String name);
    public ProductResponsedto adjustStock(Long id, StockAdjustmentRequest request);
    public ProductResponsedto updateProductPrice(Long id, PriceUpdateRequestdto request);
}

File: IMS_frontend/IMS/src/app/core/services/product.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ProductRequestDto, ProductResponseDto, StockAdjustmentRequest } from '../../shared/models/interfaces';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private readonly apiUrl = 'http://localhost:8083/api/v1/products';

  constructor(private http: HttpClient) {}

  addProduct(dto: ProductRequestDto): Observable<ProductResponseDto> {
    return this.http.post<ProductResponseDto>(`${this.apiUrl}/addproduct`, dto);
  }

  getAllProducts(): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.apiUrl}/allproducts`);
  }

  getProductById(id: number): Observable<ProductResponseDto> {
    return this.http.get<ProductResponseDto>(`${this.apiUrl}/${id}`);
  }

  getActiveProducts(): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.apiUrl}/activeproducts`);
  }

  getProductByName(name: string): Observable<ProductResponseDto> {
    return this.http.get<ProductResponseDto>(`${this.apiUrl}/productname/${name}`);
  }

  getProductBySku(sku: string): Observable<ProductResponseDto> {
    return this.http.get<ProductResponseDto>(`${this.apiUrl}/productsku/${sku}`);
  }

  getProductsByCategory(categoryName: string): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.apiUrl}/productcategory/${categoryName}`);
  }

  updateProduct(id: number, dto: ProductRequestDto): Observable<ProductResponseDto> {
    return this.http.put<ProductResponseDto>(`${this.apiUrl}/updateproduct/${id}`, dto);
  }

  deleteProduct(id: number): Observable<ProductResponseDto> {
    return this.http.delete<ProductResponseDto>(`${this.apiUrl}/deleteproduct/${id}`);
  }

  adjustStock(id: number, request: StockAdjustmentRequest): Observable<ProductResponseDto> {
    return this.http.post<ProductResponseDto>(`${this.apiUrl}/${id}/adjust-stock`, request);
  }

  getProductsBySupplier(supplierId: number): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.apiUrl}/supplier/${supplierId}`);
  }
}
